README.md
# ember-string-ishtmlsafe-polyfill [![Build Status](https://travis-ci.org/workmanw/ember-string-ishtmlsafe-polyfill.svg?branch=master)](https://travis-ci.org/workmanw/ember-string-ishtmlsafe-polyfill)

This provides a polyfill for the `Ember.String.isHTMLSafe` feature added in Ember 2.8.

RFC: [emberjs/rfcs#139](https://github.com/emberjs/rfcs/pull/139).

PR: [emberjs/ember.js#13666](https://github.com/emberjs/ember.js/pull/13666).

API: [Ember.String.htmlSafe](http://emberjs.com/api/classes/Ember.String.html#method_isHTMLSafe)


## Installation

```
ember install ember-string-ishtmlsafe-polyfill
```


## Usage

```javascript
import Ember from 'ember';

var plainString = 'plain string';
var safeString = Ember.String.htmlSafe('<div>someValue</div>');

Ember.String.isHTMLSafe(plainString); // false
Ember.String.isHTMLSafe(safeString);  // true
```


## Migration

### Applications

After you upgrade your application to Ember 2.8, you should remove `ember-string-ishtmlsafe-polyfill`
from your `package.json`.

### Addons

Addons generally support many different Ember versions, so leaving `ember-string-ishtmlsafe-polyfill`
in place for consumers of your addon is perfectly normal.  When the addon no longer supports
Ember versions older than 2.8, we recommend removing `ember-string-ishtmlsafe-polyfill` from
your `package.json` and doing a major version bump.


## Compatibility

This addon is tested against each minor Ember version starting with 1.10 through 2.8
(when `isHTMLSafe()` landed). Additionally tested against, `ember-stable`, `ember-beta`,
and `ember-canary`. A complete list can be found in the
[`ember-try.js`](https://github.com/workmanw/ember-string-ishtmlsafe-polyfill/blob/master/config/ember-try.js)
config.


## Running Tests

* `npm test` (Runs `ember try:each` to test your addon against multiple Ember versions)
* `ember test`
* `ember test --server`


CHANGELOG.md
# ember-string-ishtmlsafe Changelog

### 2.0.2 (May 28th, 2018)

- [#13](https://github.com/workmanw/ember-string-ishtmlsafe-polyfill/pull/13) Upgraded Ember-CLI to remove deprecation warnings (@oskarrough)

### 2.0.1 (April 4th, 2018)

- [#10](https://github.com/workmanw/ember-string-ishtmlsafe-polyfill/pull/10) Bumped major version of ember-cli-dependency-checker to remove deprecation warning (@bartocc)
- [#9](https://github.com/workmanw/ember-string-ishtmlsafe-polyfill/pull/9) Change to use Chrome for CI testing (@oskarrough)

### 2.0.0 (August 8th, 2017)

- [#6](https://github.com/workmanw/ember-string-ishtmlsafe-polyfill/pull/6) Removed old deprecation and ember-cli-babel dependency. (@Turbo87)
- **NOTE:** Major version bump because of the deprecation removal.

### 1.1.0 (March 3rd, 2017)

- [#2](https://github.com/workmanw/ember-string-ishtmlsafe-polyfill/pull/2) Converted module import to "native" polyfill. Also updated ember-cli and npm packages (@Panman8201)

### 1.0.1 (July 15th, 2016)

- Changed `isHtmlSafe` to `isHTMLSafe` (@workmanw)

### 1.0.0 (June 19th, 2016)

- Initial implementation (@workmanw)


tests/acceptance/initializer-test.js
import Ember from 'ember';
import { module, test } from 'qunit';
import startApp from '../../tests/helpers/start-app';
import { isHtmlSafe } from 'ember-string-ishtmlsafe/utils/string';

module('Acceptance | initializer', {
  beforeEach: function() {
    this.application = startApp();
  },

  afterEach: function() {
    Ember.run(this.application, 'destroy');
  }
});

test('Ensure `isHtmlSafe` is tacked onto `Ember.String`', function(assert) {
  visit('/');

  andThen(function() {
    assert.equal(Ember.String.isHtmlSafe, isHtmlSafe);
  });
});


tests/unit/string-ishtmlsafe-test.js
import { module, test } from 'qunit';
import Ember from 'ember';

module('Unit | isHTMLSafe');

test('"Native" Ember.String.isHTMLSafe() function properly detects safe strings', function(assert) {
  let plainStr = 'Hello World',
      htmlSafeStr = Ember.String.htmlSafe(plainStr);

  assert.notOk(Ember.String.isHTMLSafe(plainStr), 'Plain string is not html safe string');
  assert.ok(Ember.String.isHTMLSafe(htmlSafeStr), 'Html safe string matches');

  assert.notOk(Ember.String.isHTMLSafe([]), 'Array is not html safe string');
  assert.notOk(Ember.String.isHTMLSafe({}), 'Object is not html safe string');
  assert.notOk(Ember.String.isHTMLSafe(10), 'Number is not html safe string');
  assert.notOk(Ember.String.isHTMLSafe(null), 'Null is not html safe string');
});


index.js
'use strict';

var VersionChecker = require('ember-cli-version-checker');
var hasBeenWarned = false;

module.exports = {
  name: 'ember-string-ishtmlsafe-polyfill',
  included: function() {
    this._super.included.apply(this, arguments);

    this._ensureThisImport();

    var checker = new VersionChecker(this);
    var emberVersion = checker.forEmber();

    if (emberVersion.lt('2.8.0')) {
      this.import('vendor/install-ishtmlsafe-polyfill.js');
    } else if (this.parent === this.project && !hasBeenWarned){
      console.warn('ember-string-ishtmlsafe-polyfill is not required for Ember 2.8.0 and later, please remove from your `package.json`.');
      hasBeenWarned = true;
    }
  },

  _ensureThisImport: function() {
    if (!this.import) {
      this._findHost = function findHostShim() {
        var current = this;
        var app;
        do {
          app = current.app || app;
        } while (current.parent.parent && (current = current.parent));
        return app;
      };
      this.import = function importShim(asset, options) {
        var app = this._findHost();
        app.import(asset, options);
      };
    }
  }
};


.eslintrc.js
module.exports = {
  root: true,
  parserOptions: {
    ecmaVersion: 2017,
    sourceType: 'module'
  },
  plugins: [
    'ember'
  ],
  extends: [
    'eslint:recommended',
    'plugin:ember/recommended'
  ],
  env: {
    browser: true
  },
  rules: {
    'no-console': 'warn'
  },
  overrides: [
    // node files
    {
      files: [
        'index.js',
        'testem.js',
        'ember-cli-build.js',
        'config/**/*.js',
        'tests/dummy/config/**/*.js'
      ],
      excludedFiles: [
        'app/**',
        'addon/**',
        'tests/dummy/app/**'
      ],
      parserOptions: {
        sourceType: 'script',
        ecmaVersion: 2015
      },
      env: {
        browser: false,
        node: true
      },
      plugins: ['node'],
      rules: Object.assign({}, require('eslint-plugin-node').configs.recommended.rules, {
        // add your custom rules and overrides for node files here
      })
    },

    // test files
    {
      files: ['tests/**/*.js'],
      excludedFiles: ['tests/dummy/**/*.js'],
      env: {
        embertest: true
      },
      rules: {
        'ember/new-module-imports': 'warn'
      }
    }
  ]
};


config/ember-try.js
module.exports = {
  useYarn: true,
  scenarios: [
    {
      name: 'ember-1.10',
      bower: {
        dependencies: {
          'ember': '~1.10.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-1.11',
      bower: {
        dependencies: {
          'ember': '~1.11.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-1.12',
      bower: {
        dependencies: {
          'ember': '~1.12.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-1.13',
      bower: {
        dependencies: {
          'ember': '~1.13.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-2.0',
      bower: {
        dependencies: {
          'ember': '~2.0.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-2.1',
      bower: {
        dependencies: {
          'ember': '~2.1.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-2.2',
      bower: {
        dependencies: {
          'ember': '~2.2.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-2.3',
      bower: {
        dependencies: {
          'ember': '~2.3.0'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-lts-2.4',
      bower: {
        dependencies: {
          'ember': 'components/ember#lts-2-4'
        },
        resolutions: {
          'ember': 'lts-2-4'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-lts-2.8',
      bower: {
        dependencies: {
          'ember': 'components/ember#lts-2-8'
        },
        resolutions: {
          'ember': 'lts-2-8'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-lts-2.12',
      npm: {
        devDependencies: {
          'ember-source': '~2.12.0'
        }
      }
    },
    {
      name: 'ember-lts-2.16',
      npm: {
        devDependencies: {
          'ember-source': '~2.16.0'
        }
      }
    },
    {
      name: 'ember-release',
      bower: {
        dependencies: {
          'ember': 'components/ember#release'
        },
        resolutions: {
          'ember': 'release'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-beta',
      bower: {
        dependencies: {
          'ember': 'components/ember#beta'
        },
        resolutions: {
          'ember': 'beta'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-canary',
      bower: {
        dependencies: {
          'ember': 'components/ember#canary'
        },
        resolutions: {
          'ember': 'canary'
        }
      },
      npm: {
        devDependencies: {
          'ember-source': null
        }
      }
    },
    {
      name: 'ember-default',
      npm: {
        devDependencies: {}
      }
    }
  ]
};
